node/src/components/linear_chain.rs

```diff
@@ -23,11 +25,9 @@ use crate::{
     NodeRng,
 };
 
-impl<I> From<FinalitySignature> for Event<I> {
-    fn from(fs: FinalitySignature) -> Self {
-        Event::FinalitySignatureReceived(Box::new(fs))
-    }
-}
+/// The maximum number of finality signatures from a single validator we keep in memory while
+/// waiting for their block.
+const MAX_PENDING_FINALITY_SIGNATURES_PER_VALIDATOR: usize = 1000;
```

Comment:
That's a lot! Where did you come up with this number? We can make it dynamic and cap it to the number of blocks in an era?

Reply:
I don't know if it's a lot: Finality signatures are small; keeping thousands in memory won't make a big difference, and it might be good for a joining node to keep them, so once it has caught up it can immediately trust the latest block.

If we make it dynamic and tie it to eras, we'll probably do some work that overlaps with [NDRS-671](https://casperlabs.atlassian.net/browse/NDRS-671): We'll need to get chainspec and era information into the linear chain component, and possibly put the era ID into `FinalitySignature` itself.…
I'd prefer to keep it a constant for now, since I suspect that that will be a larger change than this PR.

Reply:
That's 128kb per validator. You're right, it's not that much.

Reply:
> Finality signatures are small; keeping thousands in memory won't make a big difference, and it might be good for a joining node to keep them, so once it has caught up it can immediately trust the latest block.

For [NDRS-664] I had been assuming that each validator would store their own finality signatures for finalized blocks on disk.  They could store other validator finality signatures as well.

[NDRS-664]: https://casperlabs.atlassian.net/browse/NDRS-664

Reply:
They should be stored on disk. Andreas was just ~lazy~ didn't want to postpone the main feature he works on.